osm2streets/src/transform/classify_intersections.rs

```diff
@@ -1,85 +1,229 @@
-    }
+    // Calculate all the possible movements, (except U-turns, for now).
+    let mut connections = Vec::new();
+    // Consider all pairs of roads, from s to d. Identify them using their index in the list - which
+    // is sorted in clockwise order - so that we can compare their position later.
+    for s in 0..roads.len() {
+        for d in 0..roads.len() {
+            if s == d {
+                continue; // Ignore U-turns.
+            }
 
-    // A MultiConnection is characterised by exactly one dividing line traveling through it (the
-    // line that separates traffic in different directions) and no traffic that crosses it.
-    if roads.len() == 3 {
-        let mut _num_roads_in = 0;
-        let mut _num_roads_out = 0;
-        let mut num_roads_inout = 0;
-        for road_id in roads {
-            let is_outward = road_id.i1 == *intersection_id;
-            let road = streets.roads.get(&road_id).unwrap();
-            match road.oneway_for_driving() {
-                Some(Direction::Fwd) => {
-                    if is_outward {
-                        _num_roads_out += 1
-                    } else {
-                        _num_roads_in += 1
-                    }
-                }
+            // Calculate if it is possible to emerge from s into the intersection.
+            let src_road = roads[s];
+            if !can_drive_out_of(src_road, *intersection_id) {
+                continue;
+            }
 
-                Some(Direction::Back) => {
-                    if is_outward {
-                        _num_roads_in += 1
-                    } else {
-                        _num_roads_out += 1
-                    }
-                }
-                None => num_roads_inout += 1,
+            // Calculate if it is possible to leave the intersection into d.
+            let dst_road = roads[d];
+            if !can_drive_into(dst_road, *intersection_id) {
+                continue;
+            }
+
+            // TODO detect U-Turns that should be assumed forbidden.
+            // if src and dst are oneway and
+            // adjacent on the intersection and
+            // ordered with the "insides" touching and
+            // the angle between them is small enough.
+
+            // Check for any turn restrictions.
+            if turn_is_allowed(src_road, dst_road) {
+                connections.push((s, d));
             }
         }
+    }
 
-        if num_roads_inout == 0 {
-            // The simplest MultiConnect
-            return MultiConnection;
+    // Calculate all the collisions.
+    let mut conflicts = BTreeMap::new();
+    let mut worst_conflict = Uncontested;
+    // Compare every pair of connections. Use the order of the roads around the intersection to
+    // detect if they diverge, merge, or cross.
+    // assert!(connections is sorted) so small_con large_con makes sense.
+    let mut each_con = connections.iter();
+    while let Some(small_con) = each_con.next() {
```

Comment:
Just so I understand: conflict is symmetric, so we only need to calculate "half the matrix" of pairs, right? What do you mean by sorted / small and large connections? Literally the tuples of movement indices, ie, (1, 3) < (2, 1)?

Reply:
Yes, spot on. I'll clarify this in the "every pair of connections" comment.

Reply:
In fact, the logic doesn't depend on the sort order of `small_con`/`large_con` (it may have at one point). We also aren't doing anything with the full set of conflicts, so I'll omit storing them all here.